File: android/app/src/test/kotlin/com/poracode/app/protocol/BoundedRecoveryTapeTest.kt
package com.poracode.app.protocol

import com.poracode.app.chat.RichFollowUpQueue
import com.poracode.app.chat.RichFollowUpQueueEnvelope
import com.poracode.app.chat.RichPayloadPatch
import com.poracode.app.chat.RichPendingSteer
import com.poracode.app.chat.RichRuntimeEvent
import com.poracode.app.chat.RichThreadKey
import com.poracode.app.chat.RichThreadState
import com.poracode.app.model.ClientConnectionId
import com.poracode.app.session.replay.ReplayFixtureSupport
import com.poracode.app.session.richchat.RichChatLiveFrame
import com.poracode.app.session.richchat.RichChatLiveFrameBuffer
import kotlinx.serialization.json.JsonNull
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import org.junit.Assert.assertEquals
import org.junit.Assert.assertFalse
import org.junit.Assert.assertNull
import org.junit.Assert.assertTrue
import org.junit.Test

/**
 * Drives `protocol/remote/v3/fixtures/bounded-recovery-tape.json` through the
 * Android production recovery buffers:
 *
 * - incremental cases through [ThreadHydrationCoordinator] (full fidelity:
 *   owners, generations, exact replay seqs) and through
 *   [RichChatLiveFrameBuffer] (the same count/byte/retained-age decisions and
 *   replay content, with the tape's declared byte sizes mapped to the buffer's
 *   conservative estimate and the byte budget scaled by the same ratio);
 * - replacement cases through [RichChatLiveFrameBuffer]'s follow-up-queue
 *   replay, including the stale-target case, checking every expected replay.
 *
 * The two owner-generation incremental cases (`reopen-...-rejects-stale-owner`
 * and the second stale commit in `commit-releases-...`) are only expressible on
 * the coordinator: the frame buffer has no cross-owner generation — its
 * lifecycle is the controller's selected-lease reset — so the driver resets on
 * `begin` and skips stale commits without consuming (asserting retention).
 *
 * The Kotlin buffers use a conservative UTF-16×3 estimate, so this suite pins
 * decisions, hard bounds, and replay content rather than the tape's exact
 * retained-byte totals.
 */
class BoundedRecoveryTapeTest {
    private val tape: JsonObject = ReplayFixtureSupport.readFixtureJson("bounded-recovery-tape.json")

    private val keyA = RichThreadKey(
        ClientConnectionId("30000000-0000-4000-8000-000000000003"),
        "thread-tape",
    )
    private val keyB = RichThreadKey(
        ClientConnectionId("30000000-0000-4000-8000-000000000003"),
        "thread-tape-b",
    )

    private fun obj(value: kotlinx.serialization.json.JsonElement?) = value as JsonObject
    private fun arr(value: kotlinx.serialization.json.JsonElement?) =
        (value as kotlinx.serialization.json.JsonArray)
    private fun str(value: kotlinx.serialization.json.JsonElement?) = (value as JsonPrimitive).content
    private fun int(value: kotlinx.serialization.json.JsonElement?) = str(value).toInt()
    private fun bool(value: kotlinx.serialization.json.JsonElement?) = str(value).toBoolean()

    @Test
    fun tapePinsProtocolVersionAndAdditiveBoundary() {
        assertEquals(12, int(tape["protocolVersion"]))
        assertEquals("fixture-only-additive", str(tape["versionBoundary"]))
        val semantics = obj(tape["semantics"])
        assertTrue(
            "retained age is defined against the monotonic clock",
            str(obj(semantics["incremental"])["bounds"]).contains("monotonic"),
        )
    }

    // MARK: - Incremental: ThreadHydrationCoordinator

    @Test
    fun incrementalCasesDriveThreadHydrationCoordinator() {
        for (rawCase in arr(tape["incrementalCases"])) {
            val entry = obj(rawCase)
            val id = str(entry["id"])
            val policy = obj(entry["policy"])
            val maxBytes = int(policy["maxBytes"]).toLong()
            var now = 0L
            val coord = ThreadHydrationCoordinator(
                bounds = ThreadHydrationCoordinator.Bounds(
                    maxFrames = int(policy["maxCount"]),
                    maxBytes = maxBytes,
                    maxAgeMs = int(policy["maxAgeMs"]).toLong(),
                ),
                arrivalClockMs = { now },
            )
            val expected = obj(entry["expected"])
            val commits = arr(expected["commits"])
            val generations = mutableMapOf<String, Int>()
            val ownerThread = "thread-tape"
            var currentOwner = "owner-1"
            var commitIndex = 0
            var windowBeforeFirstCommit: Triple<List<Int>, Boolean, Long>? = null

            for (rawStep in arr(entry["steps"])) {
                val step = obj(rawStep)
                when (str(step["op"])) {
                    "begin" -> {
                        currentOwner = str(step["owner"])
                        generations[currentOwner] = coord.beginOpen(ownerThread)
                        step["expectAfter"]?.let {
                            assertEquals(id, 0, coord.bufferedCount())
                            assertFalse(id, coord.overflowed)
                        }
                    }

                    "append" -> {
                        val gen = generations.getValue(currentOwner)
                        val declaredBytes = int(step["bytes"])
                        now = maxOf(now, int(step["atMs"]).toLong())
                        // A JSON string payload whose conservative estimate is
                        // close to the tape's declared size (decisions below
                        // stay far from the byte-budget boundary).
                        val payload = JsonPrimitive("x".repeat(maxOf(0, (declaredBytes - 38) / 3)))
                        coord.bufferFrame(
                            seq = int(step["seq"]),
                            threadId = ownerThread,
                            event = payload,
                            openGeneration = gen,
                            arrivalMs = int(step["atMs"]).toLong(),
                        )
                        assertTrue(
                            "$id: retained bytes stay within the hard bound",
                            coord.bufferedBytes() <= maxBytes,
                        )
                    }

                    "advance" -> now = maxOf(now, int(step["atMs"]).toLong())

                    "reset" -> {
                        coord.cancel()
                        step["expectAfter"]?.let {
                            assertEquals(id, 0, coord.bufferedCount())
                            assertFalse(id, coord.overflowed)
                        }
                    }

                    "commit" -> {
                        if (windowBeforeFirstCommit == null) {
                            windowBeforeFirstCommit = Triple(
                                coord.bufferedSeqs(),
                                coord.overflowed,
                                coord.bufferedBytes(),
                            )
                        }
                        val expectedCommit = obj(commits[commitIndex])
                        commitIndex += 1
                        val generation = generations.getValue(str(step["owner"]))
                        val completion = coord.completeHistory(
                            threadId = ownerThread,
                            openGeneration = generation,
                            snapshotSeq = int(step["snapshotSeq"]),
                        )
                        when (str(expectedCommit["outcome"])) {
                            "replayed" -> {
                                val replayed = requireNotNull(completion) { "$id: expected replay" }
                                assertEquals(
                                    id,
                                    arr(expectedCommit["replaySeqs"]).map { int(it) },
                                    replayed.replay.map { it.seq },
                                )
                                assertEquals(
                                    "$id: coverage loss at commit",
                                    bool(expectedCommit["coverageLost"]),
                                    replayed.coverageLost,
                                )
                            }

                            "stale" -> assertNull("$id: stale owner must not install", completion)
                        }
                        assertEquals(
                            "$id: release after commit",
                            bool(expectedCommit["released"]),
                            coord.bufferedCount() == 0,
                        )
                    }
                }
            }
            assertEquals("$id: every expected commit is exercised", commits.size, commitIndex)
            val window = requireNotNull(windowBeforeFirstCommit) { "$id: no commit step" }
            assertEquals("$id: retained window", arr(expected["retainedSeqs"]).map { int(it) }, window.first)
            assertEquals("$id: coverage lost", bool(expected["coverageLost"]), window.second)
            assertTrue("$id: hard byte bound before commit", window.third <= maxBytes)
        }
    }

    // MARK: - Incremental: RichChatLiveFrameBuffer

    @Test
    fun incrementalCasesDriveRichChatLiveFrameBuffer() {
        for (rawCase in arr(tape["incrementalCases"])) {
            val entry = obj(rawCase)
            val id = str(entry["id"])
            val policy = obj(entry["policy"])
            val maxBytes = int(policy["maxBytes"]).toLong()
            // The tape's declared bytes are exact; the production buffer uses a
            // conservative estimate. Scale the byte budget by the estimate of a
            // 100-byte frame so the relative count/byte decisions match.
            val scaledMaxBytes = maxOf(
                1L,
                (maxBytes.toDouble() / 100.0 * estimateOf(100)).toLong(),
            )
            var now = 0L
            val buffer = RichChatLiveFrameBuffer(
                maxFrames = int(policy["maxCount"]),
                maxBytes = scaledMaxBytes,
                maxAgeMs = int(policy["maxAgeMs"]).toLong(),
                arrivalClockMs = { now },
            )
            val expected = obj(entry["expected"])
            val commits = arr(expected["commits"])
            var commitIndex = 0
            var windowBeforeFirstCommit: Pair<List<Int>, Boolean>? = null

            for (rawStep in arr(entry["steps"])) {
                val step = obj(rawStep)
                when (str(step["op"])) {
                    "begin" -> {
                        buffer.reset()
                        step["expectAfter"]?.let {
                            assertEquals(id, 0, buffer.bufferedCount())
                            assertFalse(id, buffer.overflow)
                        }
                    }

                    "append" -> {
                        now = maxOf(now, int(step["atMs"]).toLong())
                        buffer.buffer(frameFor(int(step["seq"]), int(step["bytes"])))
                        assertTrue(
                            "$id: retained bytes stay within the hard bound",
                            buffer.bufferedEstimatedBytes() <= scaledMaxBytes,
                        )
                    }

                    "advance" -> now = maxOf(now, int(step["atMs"]).toLong())

                    "reset" -> {
                        buffer.reset()
                        step["expectAfter"]?.let {
                            assertEquals(id, 0, buffer.bufferedCount())
                            assertFalse(id, buffer.overflow)
                        }
                    }

                    "commit" -> {
                        if (windowBeforeFirstCommit == null) {
                            windowBeforeFirstCommit =
                                buffer.bufferedSequences().filterNotNull() to buffer.overflow
                        }
                        val expectedCommit = obj(commits[commitIndex])
                        commitIndex += 1
                        val snapshotSeq = int(step["snapshotSeq"])
                        when (str(expectedCommit["outcome"])) {
                            "replayed" -> {
                                val replayed = buffer.replayAfterSnapshot(snapshot(seq = snapshotSeq))
                                assertEquals(
                                    "$id: replay content",
                                    arr(expectedCommit["replaySeqs"]).map { int(it) }.sorted(),
                                    replayed.transcript.itemsById.keys
                                        .map { it.removePrefix("item-").toInt() }
                                        .sorted(),
                                )
                                assertEquals(
                                    "$id: coverage loss at commit",
                                    bool(expectedCommit["coverageLost"]),
                                    replayed.hadOverflow,
                                )
                            }

                            // Owner mismatch has no frame-buffer equivalent: a
                            // stale commit must not consume the retained window.
                            "stale" -> Unit
                        }
                        assertEquals(
                            "$id: release after commit",
                            bool(expectedCommit["released"]),
                            buffer.bufferedCount() == 0,
                        )
                    }
                }
            }
            assertEquals("$id: every expected commit is exercised", commits.size, commitIndex)
            val window = requireNotNull(windowBeforeFirstCommit) { "$id: no commit step" }
            assertEquals("$id: retained window", arr(expected["retainedSeqs"]).map { int(it) }, window.first)
            assertEquals("$id: coverage lost", bool(expected["coverageLost"]), window.second)
        }
    }

    /** Estimate of one frame carrying the tape's declared byte payload. */
    private fun estimateOf(declaredBytes: Int): Long {
        val buffer = RichChatLiveFrameBuffer(
            maxFrames = 8,
            maxBytes = Long.MAX_VALUE / 2,
            maxAgeMs = Long.MAX_VALUE / 2,
            arrivalClockMs = { 0L },
        )
        buffer.buffer(frameFor(seq = 1, declaredBytes = declaredBytes))
        return buffer.bufferedEstimatedBytes()
    }

    private fun frameFor(seq: Int, declaredBytes: Int): RichChatLiveFrame {
        val payloadLength = maxOf(0, (declaredBytes - 64) / 3)
        return RichChatLiveFrame(
            sequence = seq,
            events = listOf(
                RichRuntimeEvent.ItemStarted(
                    threadKey = keyA,
                    itemId = "item-$seq",
                    itemType = "assistant_message",
                    payload = RichPayloadPatch.Absent,
                    parentItemId = null,
                ),
                RichRuntimeEvent.ContentDelta(
                    threadKey = keyA,
                    itemId = "item-$seq",
                    stream = "assistant_text",
                    delta = "x".repeat(payloadLength),
                ),
            ),
        )
    }

    // MARK: - Replacement

    @Test
    fun replacementCasesDriveFollowUpQueueFrameReplay() {
        for (rawCase in arr(tape["replacementCases"])) {
            val entry = obj(rawCase)
            val id = str(entry["id"])
            val expected = obj(entry["expected"])
            val records = arr(entry["records"])
            val recordTarget = records.firstOrNull()?.let { obj(it)["target"] }?.let { str(it) }
                ?: "target-1"
            val recordKey = if (recordTarget == "target-2") keyB else keyA

            // Window probe: replay at snapshotSeq -1 so a retained state always
            // wins and the retained value is observable without consuming the
            // buffer used for the expected replay sequence.
            val probe = RichChatLiveFrameBuffer(
                maxFrames = 8,
                maxBytes = 1_000_000,
                maxAgeMs = 600_000,
                arrivalClockMs = { 0L },
            )
            recordReplacement(probe, entry, records, recordKey)
            assertFalse("$id: supersede never loses coverage", probe.overflow)
            val probeBase = RichThreadState.hydrate(recordKey, emptyList())
                .copy(followUpQueue = queue("queue-snapshot"))
            val probeResult = probe.replayAfterSnapshot(
                snapshot(seq = -1, key = recordKey, base = probeBase),
                probeBase,
            )
            val probeState = probeResult.transcript.followUpQueue?.items?.firstOrNull()?.id
            // The frame buffer expresses the replacement policy through the
            // replay result (replace-on-event, newest wins), not a one-slot
            // count: the probe at snapshotSeq -1 must show the tape's retained
            // state, and a reset slot must leave the base untouched.
            val retainedState = expected["retainedState"]
            if (retainedState is JsonNull) {
                assertEquals("$id: released slot retains nothing", 0, probe.bufferedCount())
                assertEquals("$id: empty slot replays the base", "queue-snapshot", probeState)
            } else {
                assertEquals("$id: retained state", str(retainedState), probeState)
            }

            // Every expected replay is exercised against a fresh buffer.
            val buffer = RichChatLiveFrameBuffer(
                maxFrames = 8,
                maxBytes = 1_000_000,
                maxAgeMs = 600_000,
                arrivalClockMs = { 0L },
            )
            recordReplacement(buffer, entry, records, recordKey)
            assertFalse("$id: supersede never loses coverage", buffer.overflow)
            val replayTarget = entry["replayTarget"]?.let { str(it) } ?: "target-1"
            val replayKey = if (replayTarget == "target-2") keyB else keyA
            for (rawReplay in arr(expected["replays"])) {
                val replay = obj(rawReplay)
                val base = RichThreadState.hydrate(replayKey, emptyList())
                    .copy(followUpQueue = queue(str(replay["over"])))
                val result = buffer.replayAfterSnapshot(
                    snapshot(seq = int(replay["snapshotSeq"]), key = replayKey, base = base),
                    base,
                )
                assertEquals(
                    "$id: replay result",
                    str(replay["state"]),
                    result.transcript.followUpQueue?.items?.firstOrNull()?.id,
                )
                assertFalse("$id: replay must not demand recovery", result.hadOverflow)
            }
            assertEquals("$id: replays release the window", 0, buffer.bufferedCount())
        }
    }

    private fun recordReplacement(
        buffer: RichChatLiveFrameBuffer,
        entry: JsonObject,
        records: List<kotlinx.serialization.json.JsonElement>,
        recordKey: RichThreadKey,
    ) {
        for (rawRecord in records) {
            val record = obj(rawRecord)
            buffer.buffer(
                RichChatLiveFrame(
                    sequence = int(record["sequence"]),
                    events = emptyList(),
                    followUpQueue = RichFollowUpQueueEnvelope(
                        threadKey = recordKey,
                        queue = queue(str(record["state"])),
                    ),
                ),
            )
        }
        if (entry["reset"]?.let { bool(it) } == true) buffer.reset()
    }

    private fun queue(state: String) = RichFollowUpQueue(
        items = listOf(RichPendingSteer(id = state, prompt = state, stagedAtEpochMs = 0.0)),
        paused = false,
    )

    private fun snapshot(
        seq: Int,
        key: RichThreadKey = keyA,
        base: RichThreadState = RichThreadState.hydrate(key, emptyList()),
    ) = com.poracode.app.session.richchat.RichChatHistorySnapshot(
        key = key,
        snapshotSeq = seq,
        state = base,
        olderCursor = null,
        config = com.poracode.app.model.ThreadConfig(model = "gpt-5"),
        terminalScrollback = null,
        updatedAt = "2026-08-12T00:00:00.000Z",
    )
}

File: protocol/remote/v3/bounded-recovery-tape.test.ts
import { readFileSync } from "node:fs";
import { describe, expect, it } from "vitest";
import { z } from "zod";
import { PORACODE_REMOTE_PROTOCOL_VERSION } from "../../../src/shared/remote/protocol";

/**
 * Shared bounded-recovery semantics. This suite is the reference evaluator for
 * `fixtures/bounded-recovery-tape.json`; the Swift and Kotlin suites drive the
 * same tape through the production recovery buffers. The tape is client-local
 * bookkeeping (count/byte/age bounds, overflow recovery, replacement
 * supersede), so there is no wire shape or generated artifact to bump.
 */

const policySchema = z.object({
  maxCount: z.number().int().positive(),
  maxBytes: z.number().int().positive(),
  maxAgeMs: z.number().int().nonnegative(),
});

const retainedSchema = z.object({
  retainedSeqs: z.array(z.number().int()),
  retainedBytes: z.number().int().nonnegative(),
  coverageLost: z.boolean(),
});

const appendStepSchema = z.object({
  op: z.literal("append"),
  seq: z.number().int(),
  bytes: z.number().int().nonnegative(),
  atMs: z.number().int().nonnegative(),
});
const beginStepSchema = z.object({ op: z.literal("begin"), owner: z.string().min(1) });
const commitStepSchema = z.object({
  op: z.literal("commit"),
  owner: z.string().min(1),
  snapshotSeq: z.number().int(),
});
const resetStepSchema = z.object({
  op: z.literal("reset"),
  expectAfter: retainedSchema,
});
const advanceStepSchema = z.object({
  op: z.literal("advance"),
  atMs: z.number().int().nonnegative(),
});
const stepSchema = z.discriminatedUnion("op", [
  appendStepSchema,
  beginStepSchema,
  commitStepSchema,
  resetStepSchema,
  advanceStepSchema,
]);

const incrementalCaseSchema = z.object({
  id: z.string().min(1),
  policy: policySchema,
  steps: z.array(stepSchema).min(1),
  expected: retainedSchema.extend({
    commits: z.array(
      z.object({
        owner: z.string().min(1),
        snapshotSeq: z.number().int(),
        outcome: z.enum(["replayed", "stale"]),
        replaySeqs: z.array(z.number().int()),
        coverageLost: z.boolean(),
        released: z.boolean(),
      }),
    ),
  }),
});

const replacementCaseSchema = z.object({
  id: z.string().min(1),
  records: z.array(
    z.object({
      sequence: z.number().int(),
      state: z.string().min(1),
      target: z.string().min(1).optional(),
    }),
  ),
  reset: z.boolean().optional(),
  replayTarget: z.string().min(1).optional(),
  expected: z.object({
    retainedCount: z.number().int().nonnegative(),
    retainedSequence: z.number().int().nullable(),
    retainedState: z.string().min(1).nullable(),
    coverageLost: z.literal(false),
    replays: z.array(
      z.object({
        over: z.string().min(1),
        snapshotSeq: z.number().int(),
        state: z.string().min(1),
      }),
    ),
  }),
});

const tapeSchema = z.object({
  id: z.literal("remote-v3-bounded-recovery-tape"),
  format: z.literal("remote-v3-bounded-recovery-tape"),
  protocolVersion: z.number().int(),
  versionBoundary: z.literal("fixture-only-additive"),
  semantics: z.object({
    incremental: z.record(z.string(), z.string()),
    replacement: z.record(z.string(), z.string()),
  }),
  incrementalCases: z.array(incrementalCaseSchema).min(1),
  replacementCases: z.array(replacementCaseSchema).min(1),
});

type AppendStep = z.infer<typeof appendStepSchema>;
type BeginStep = z.infer<typeof beginStepSchema>;
type CommitStep = z.infer<typeof commitStepSchema>;
type ResetStep = z.infer<typeof resetStepSchema>;
type AdvanceStep = z.infer<typeof advanceStepSchema>;
type IncrementalCase = z.infer<typeof incrementalCaseSchema>;
type ReplacementCase = z.infer<typeof replacementCaseSchema>;

const tape = tapeSchema.parse(
  JSON.parse(
    readFileSync(new URL("./fixtures/bounded-recovery-tape.json", import.meta.url), "utf8"),
  ),
);

interface IncrementalFrame {
  readonly seq: number;
  readonly bytes: number;
  readonly atMs: number;
}

/**
 * Reference bounded incremental recovery buffer. Hard count/byte/retained-age
 * bounds against one monotonic clock, oldest-first eviction, and coverage loss
 * on any eviction or expiry. Arrivals and the clock share the tape's `atMs`
 * base: appends advance the clock to their arrival, `advance` steps move it
 * with no new input, and commit drops any retained frame that outlived
 * `maxAgeMs` before replaying.
 */
class ReferenceIncrementalBuffer {
  private frames: IncrementalFrame[] = [];
  private activeOwner: string | null = null;
  private clockMs = 0;
  coverageLost = false;

  constructor(private readonly policy: z.infer<typeof policySchema>) {}

  private byteCount(): number {
    return this.frames.reduce((total, frame) => total + frame.bytes, 0);
  }

  private expired(frame: IncrementalFrame): boolean {
    return this.clockMs - frame.atMs > this.policy.maxAgeMs;
  }

  begin(owner: string): void {
    this.activeOwner = owner;
    this.frames = [];
    this.coverageLost = false;
  }

  advance(atMs: number): void {
    this.clockMs = Math.max(this.clockMs, atMs);
  }

  append(seq: number, bytes: number, atMs: number): boolean {
    if (this.activeOwner === null) return false;
    this.advance(atMs);
    this.frames.push({ seq, bytes, atMs });
    for (;;) {
      const oldest = this.frames.at(0);
      const overCount = this.frames.length > this.policy.maxCount;
      const overBytes = this.byteCount() > this.policy.maxBytes;
      const overAge = oldest !== undefined && this.expired(oldest);
      if (!overCount && !overBytes && !overAge) break;
      this.frames.shift();
      this.coverageLost = true;
    }
    return true;
  }

  reset(): void {
    this.activeOwner = null;
    this.frames = [];
    this.coverageLost = false;
  }

  retained(): { retainedSeqs: number[]; retainedBytes: number; coverageLost: boolean } {
    return {
      retainedSeqs: this.frames.map((frame) => frame.seq),
      retainedBytes: this.byteCount(),
      coverageLost: this.coverageLost,
    };
  }

  commit(owner: string, snapshotSeq: number) {
    if (this.activeOwner !== owner) {
      return {
        outcome: "stale" as const,
        replaySeqs: [] as number[],
        coverageLost: this.coverageLost,
        released: this.frames.length === 0,
      };
    }
    // Expiry is evaluated again at commit: a hung or quiet read releases its
    // payload here and demands recovery instead of replaying stale frames.
    const expired = this.frames.filter((frame) => this.expired(frame));
    if (expired.length > 0) {
      const expiredSeqs = new Set(expired.map((frame) => frame.seq));
      this.frames = this.frames.filter((frame) => !expiredSeqs.has(frame.seq));
      this.coverageLost = true;
    }
    const coverageLost = this.coverageLost;
    const replaySeqs = this.frames
      .filter((frame) => frame.seq > snapshotSeq)
      .map((frame) => frame.seq);
    this.reset();
    return { outcome: "replayed" as const, replaySeqs, coverageLost, released: true };
  }
}

class ReferenceReplacementSlot {
  private retained: { sequence: number; state: string; target: string } | null = null;
  coverageLost = false;

  record(sequence: number, state: string, target: string): void {
    if (this.retained === null || sequence >= this.retained.sequence) {
      this.retained = { sequence, state, target };
    }
  }

  replay(over: string, snapshotSeq: number, target: string): string {
    if (this.retained === null) return over;
    if (this.retained.target !== target) return over;
    return this.retained.sequence > snapshotSeq ? this.retained.state : over;
  }

  reset(): void {
    this.retained = null;
  }

  retainedState(): {
    retainedCount: number;
    retainedSequence: number | null;
    retainedState: string | null;
  } {
    return {
      retainedCount: this.retained === null ? 0 : 1,
      retainedSequence: this.retained?.sequence ?? null,
      retainedState: this.retained?.state ?? null,
    };
  }
}

function assertAppend(buffer: ReferenceIncrementalBuffer, step: AppendStep): void {
  expect(buffer.append(step.seq, step.bytes, step.atMs)).toBe(true);
}

function assertBegin(buffer: ReferenceIncrementalBuffer, step: BeginStep): void {
  buffer.begin(step.owner);
}

function assertReset(buffer: ReferenceIncrementalBuffer, step: ResetStep): void {
  buffer.reset();
  expect(buffer.retained()).toEqual(step.expectAfter);
}

function assertAdvance(buffer: ReferenceIncrementalBuffer, step: AdvanceStep): void {
  buffer.advance(step.atMs);
}

function assertCommit(
  buffer: ReferenceIncrementalBuffer,
  entry: IncrementalCase,
  step: CommitStep,
  commitIndex: number,
): number {
  const expected = entry.expected.commits[commitIndex];
  if (expected === undefined)
    throw new Error(`${entry.id}: missing expected commit ${commitIndex}`);
  expect(expected.owner).toBe(step.owner);
  expect(expected.snapshotSeq).toBe(step.snapshotSeq);
  expect(buffer.commit(step.owner, step.snapshotSeq)).toEqual({
    outcome: expected.outcome,
    replaySeqs: expected.replaySeqs,
    coverageLost: expected.coverageLost,
    released: expected.released,
  });
  return commitIndex + 1;
}

function evaluateIncrementalCase(entry: IncrementalCase): void {
  const buffer = new ReferenceIncrementalBuffer(entry.policy);
  let commitIndex = 0;
  let windowBeforeFirstCommit: ReturnType<ReferenceIncrementalBuffer["retained"]> | null = null;
  for (const step of entry.steps) {
    switch (step.op) {
      case "begin":
        assertBegin(buffer, step);
        break;
      case "append":
        assertAppend(buffer, step);
        break;
      case "reset":
        assertReset(buffer, step);
        break;
      case "advance":
        assertAdvance(buffer, step);
        break;
      case "commit":
        windowBeforeFirstCommit ??= buffer.retained();
        commitIndex = assertCommit(buffer, entry, step, commitIndex);
        break;
    }
  }
  expect(commitIndex).toBe(entry.expected.commits.length);
  expect(windowBeforeFirstCommit).toEqual({
    retainedSeqs: entry.expected.retainedSeqs,
    retainedBytes: entry.expected.retainedBytes,
    coverageLost: entry.expected.coverageLost,
  });
}

function assertReplay(
  slot: ReferenceReplacementSlot,
  entry: ReplacementCase,
  replay: ReplacementCase["expected"]["replays"][number],
): void {
  const target = entry.replayTarget ?? "target-1";
  expect(slot.replay(replay.over, replay.snapshotSeq, target)).toBe(replay.state);
}

function evaluateReplacementCase(entry: ReplacementCase): void {
  const slot = new ReferenceReplacementSlot();
  for (const record of entry.records) {
    slot.record(record.sequence, record.state, record.target ?? "target-1");
  }
  if (entry.reset === true) slot.reset();
  expect({
    ...slot.retainedState(),
    coverageLost: slot.coverageLost,
  }).toEqual({
    retainedCount: entry.expected.retainedCount,
    retainedSequence: entry.expected.retainedSequence,
    retainedState: entry.expected.retainedState,
    coverageLost: entry.expected.coverageLost,
  });
  entry.expected.replays.forEach((replay) => {
    assertReplay(slot, entry, replay);
  });
}

describe("bounded-recovery tape (shared client-local semantics)", () => {
  it("pins the protocol version and additive fixture boundary", () => {
    expect(tape.protocolVersion).toBe(PORACODE_REMOTE_PROTOCOL_VERSION);
    expect(tape.semantics.incremental["overflow"]).toContain("authoritative recovery");
    expect(tape.semantics.incremental["bounds"]).toContain("monotonic");
    expect(tape.semantics.incremental["age"]).toContain("commit");
    expect(tape.semantics.replacement["overflow"]).toContain("never");
  });

  it.each(tape.incrementalCases.map((entry) => [entry.id, entry] as const))(
    "incremental: %s",
    (_id, entry) => {
      // Every incremental case must pin at least one commit outcome.
      expect(entry.expected.commits.length).toBeGreaterThan(0);
      evaluateIncrementalCase(entry);
    },
  );

  it.each(tape.replacementCases.map((entry) => [entry.id, entry] as const))(
    "replacement: %s",
    (_id, entry) => {
      // Every replacement case must record at least one state.
      expect(entry.records.length).toBeGreaterThan(0);
      evaluateReplacementCase(entry);
    },
  );
});
